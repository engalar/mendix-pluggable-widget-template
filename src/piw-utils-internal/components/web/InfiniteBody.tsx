import { ReactElement, useCallback, useState, useRef, useLayoutEffect, PropsWithChildren, CSSProperties } from "react";
import classNames from "classnames";

interface InfiniteBodyProps {
    className?: string;
    hasMoreItems: boolean;
    isInfinite: boolean;
    role?: string;
    setPage?: (computePage: (prevPage: number) => number) => void;
    style?: CSSProperties;
}

export function InfiniteBody(props: PropsWithChildren<InfiniteBodyProps>): ReactElement {
    const [bodySize, setBodySize] = useState(0);
    const containerRef = useRef<HTMLDivElement>(null);

    const trackScrolling = useCallback(
        e => {
            /**
             * In Windows OS the result of first expression returns a non integer and result in never loading more, require floor to solve.
             */
            const bottom = Math.floor(e.target.scrollHeight - e.target.scrollTop) === Math.floor(e.target.clientHeight);
            if (bottom) {
                if (props.hasMoreItems && props.setPage) {
                    props.setPage(prev => prev + 1);
                }
            }
        },
        [props.hasMoreItems, props.setPage]
    );

    const calculateBodyHeight = useCallback((): void => {
        if (props.isInfinite && props.hasMoreItems && bodySize <= 0 && containerRef.current) {
            setBodySize(containerRef.current.clientHeight - 30);
        }
    }, [props.isInfinite, props.hasMoreItems, bodySize]);

    useLayoutEffect(() => {
        setTimeout(() => calculateBodyHeight(), 100);
    }, [calculateBodyHeight]);

    return (
        <div
            ref={containerRef}
            className={classNames(props.className, props.isInfinite ? "infinite-loading" : "")}
            onScroll={props.isInfinite ? trackScrolling : undefined}
            role={props.role}
            style={props.isInfinite && bodySize > 0 ? { ...props.style, maxHeight: bodySize } : props.style}
        >
            {props.children}
        </div>
    );
}
