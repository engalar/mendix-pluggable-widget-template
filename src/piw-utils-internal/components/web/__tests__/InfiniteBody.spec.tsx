import { render, mount } from "enzyme";
import { InfiniteBody } from "../InfiniteBody";

describe("Infinite Body", () => {
    it("renders the structure correctly", () => {
        const component = render(<InfiniteBody isInfinite setPage={jest.fn()} hasMoreItems />);

        expect(component).toMatchSnapshot();
    });

    it("triggers correct events on scroll", () => {
        const setPage = jest.fn();
        const component = mount(<InfiniteBody isInfinite setPage={setPage} hasMoreItems style={{ maxHeight: 50 }} />);

        component.find(".infinite-loading").simulate("scroll", {
            target: {
                scrollHeight: 200,
                scrollTop: 100,
                clientHeight: 100
            }
        });

        expect(setPage).toBeCalled();
    });
});
