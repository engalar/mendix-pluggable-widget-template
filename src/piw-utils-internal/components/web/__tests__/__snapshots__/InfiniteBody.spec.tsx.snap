// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infinite Body renders the structure correctly 1`] = `
<div
  class="infinite-loading"
/>
`;
